lib/srv/reexec.go

```diff
@@ -740,63 +573,380 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+
+	// Networking resources created from in main PAM thread should also be cleaned up
+	// in the main PAM thread. Cleanup will either happen when the networking resources's
+	// connection to the parent is closed, or when this process exits (deferred).
+	cleanupC := make(chan networkingResource)
+	defer func() {
+		go func() {
+			wg.Wait()
+			close(cleanupC)
+		}()
+
+		for netResource := range cleanupC {
+			netResource.Close()
+		}
+	}()
+
+	// Start the main request goroutine.
+	go func() {
+		for {
+			buf := make([]byte, networking.RequestBufferSize)
+			fbuf := make([]*os.File, 1)
+			n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
+			if err != nil {
+				// parent connection closed by interrupt signal.
+				if utils.IsOKNetworkError(err) {
+					return
+				}
+				writeErrorToConn(parentConn, err)
+				continue
+			}
+
+			if fn == 0 {
+				writeErrorToConn(parentConn, trace.BadParameter("networking request missing per-request uds file"))
+				continue
+			}
+
+			req := new(networkingRequest)
+			req.conn, err = uds.FromFile(fbuf[0])
+			_ = fbuf[0].Close()
+			if err != nil {
+				writeErrorToConn(parentConn, trace.BadParameter("failed to get a connection from control file"))
+				continue
+			}
+			context.AfterFunc(ctx, func() { req.conn.Close() })
+
+			if err := json.Unmarshal(buf[:n], &req.Request); err != nil {
+				writeErrorToConn(parentConn, trace.Wrap(err, "failed to parse networking request"))
+				_ = req.conn.Close()
+				continue
+			}
+
+			switch req.Operation {
+			case networking.NetworkingOperationDial, networking.NetworkingOperationListen:
+				switch req.Network {
+				case "tcp":
+					// There are currently no known issues with tcp listen/dial in a multithreaded PAM context.
+					// Handle networking resource creation and cleanup in a new goroutine.
+					go func() {
+						netResource, err := handleNetworkingRequest(ctx, req)
+						if err != nil {
+							writeErrorToConn(req.conn, err)
+							return
+						}
+
+						wg.Add(1)
+						req.onClose(func() {
+							defer wg.Done()
+							netResource.Close()
+						})
+					}()
+				default:
+					// Note: we don't currently support non-tcp network forwarding, so this branch is not
+					// currently reached. If in the future we add unix socket forwarding similar to OpenSSH's
+					// direct-streamlocal@openssh.com extension, we should revisit this multithreading limitation
+					// to prevent performance degradation.
+					requestC <- req
+				}
+			case networking.NetworkingOperationListenAgent, networking.NetworkingOperationListenX11:
+				// Agent and X11 forwarding requests should occur very rarely, so handling
+				// them in the main thread should have negligible performance impact.
+				requestC <- req
+			}
+		}
+	}()
+
+	// Some requests and cleanup must happen in the main PAM thread. Handle any requests
+	// or errors passed from the main request goroutine above, post connection cleanup,
+	// or interrupt signals from the parent process.
```

Comment:
I would make this a bit more generic and have the goroutine (tied to the PAM thread) range over a `chan func()` that's "owned" by the side goroutine and execute whatever comes through, then return whatever's in some variables (as set by the side goroutine before closing the channel).